beacon-chain/p2p/fork.go

```diff
@@ -0,0 +1,127 @@
+package p2p
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/p2p/enr"
+	"github.com/prysmaticlabs/go-ssz"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/sirupsen/logrus"
+)
+
+// ENR key used for eth2-related fork data.
+const eth2EnrKey = "eth2"
+
+// EnrForkID represents a special value ssz-encoded into
+// the local node's ENR for discovery purposes. Peers should
+// only connect if their enrForkID matches.
+type EnrForkID struct {
+	CurrentForkDigest [4]byte
+	NextForkVersion   [4]byte
+	NextForkEpoch     uint64
+}
+
+// Compares fork ENRs between an incoming peer's record and our node's
+// local record values for current and next fork version/epoch.
+func (s *Service) compareForkENR(record *enr.Record) error {
+	currentRecord := s.dv5Listener.LocalNode().Node().Record()
+	peerForkENR, err := retrieveForkEntry(record)
+	if err != nil {
+		return err
+	}
+	currentForkENR, err := retrieveForkEntry(currentRecord)
+	if err != nil {
+		return err
+	}
+	// Clients SHOULD connect to peers with current_fork_digest, next_fork_version,
+	// and next_fork_epoch that match local values.
+	if peerForkENR.CurrentForkDigest != currentForkENR.CurrentForkDigest {
+		return fmt.Errorf(
+			"fork digest of peer: %v, does not match local value: %v",
+			peerForkENR.CurrentForkDigest,
+			currentForkENR.CurrentForkDigest,
+		)
+	}
+	// Clients MAY connect to peers with the same current_fork_version but a
```

Comment:
Given the language of this, should we have a flag to toggle whether to connect or not to connect?

Reply:
I think this is fine, the rest of the language of the spec says:

- Unless `ENRForkID` is manually updated to matching prior to the earlier `next_fork_epoch` of the two clients, these type of connecting clients will be unable to successfully interact starting at the earlier `next_fork_epoch`

I don't think we should worry about this too much for now